### PR TITLE
core: arm: core_mmu: fix VA overflow issue in assign_mem_va()

### PR DESCRIPTION
**Pull request by maximus64**
assign_mem_va() missing VA limit check cause TEE panic when
VA exceed the width limit. Add check to catch this case.

Signed-off-by: Khoa Hoang <admin@khoahoang.com>

### CODE REVIEW

File: core/arch/arm/mm/core_mmu.c [hunk 1]
```diff
@@ -935,6 +935,8 @@ static bool assign_mem_va(vaddr_t tee_ram_va,
 			map->va = va;
 			if (ADD_OVERFLOW(va, map->size, &va))
 				return false;
+			if (va & ~va_mask)
+				return false;
```
Comment (etienne-lms) [inline core/arch/arm/mm/core_mmu.c:939]: This is needed for 64bit VA space. I suggest to not add `va_mask` as argument to this function and use here:
```c
			if (IS_ENABLED(ARM64) && va >= BIT(get_va_width()))
				return false;
```
and same the the change below.

**Comment by etienne-lms** [comment]
I meant this change will not prevent TEE to panic, it will rather make TEE panicking at the right place.

**Comment by maximus64** [comment]
> I meant this change will not prevent TEE to panic, it will rather make TEE panicking at the right place.

Yea. But in the case with ASLR this change will help since it will fail to map at that offset and then the loop will do bit flip and retry again.

File: core/arch/arm/mm/core_mmu.c [hunk 2]
```diff
@@ -909,6 +909,15 @@ static void assign_mem_granularity(struct tee_mmap_region *memory_map)
 	}
 }
 
+static unsigned int get_va_width(void)
+{
+#ifdef ARM64
+	return 64 - __builtin_ctzl(CFG_LPAE_ADDR_SPACE_SIZE);
+#else
+	return 32;
+#endif
```
Comment (Emantor) [inline core/arch/arm/mm/core_mmu.c:918]:
```suggestion
    if (IS_ENABLED(ARM64) {
        return 64 - __builtin_ctzl(CFG_LPAE_ADDR_SPACE_SIZE);
    } else {
        return 32;
    }
```
(somehow the suggested changes interface won't let me insert tabs…)

Comment (etienne-lms) [reply core/arch/arm/mm/core_mmu.c:918]: preferably without braces
```c
static unsigned int get_va_width(void)
{
	if (IS_ENABLED(ARM64)
		return 64 - __builtin_ctzl(CFG_LPAE_ADDR_SPACE_SIZE);
	else
		return 32;
}

**Comment by jenswi-linaro** [comment]
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

**Comment by jforissier** [comment]
Ignoring the IBART error, this PR just needs rebasing onto master.

**Pull request merged**